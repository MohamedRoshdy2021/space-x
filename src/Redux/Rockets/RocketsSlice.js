import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

// Async thunk to fetch data from the API
export const fetchRockets = createAsyncThunk('rockets/fetchRockets', async () => {
  const response = await axios.get('https://api.spacexdata.com/v3/rockets');
  return response.data;
});

const rocketsSlice = createSlice({
  name: 'Rockets',
  initialState: [],
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchRockets.fulfilled, (state, action) => [...state, ...action.payload]);
  },
});

export const selectRockets = (state) => state.rockets;

export default rocketsSlice.reducer;
