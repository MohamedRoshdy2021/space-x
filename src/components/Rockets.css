.parent-div {
  display: flex;
  align-items: start;
  width: 95%;
  gap: 2rem;
  margin: 1rem auto 0;
  position: relative;
  top: 20px;
  bottom: 20px;
}

.rockets-image {
  width: 200px;
  height: 160px;
}

.text-div h2 {
  margin: 0;
  padding: 0;
}

.reserve-button {
  background-color: #2870f4;
  border: none;
  padding: 0.4rem 0.8rem;
  color: #fff;
  border-radius: 4px;
  cursor: pointer;
}

.cancel-reserve-button {
  padding: 0.4rem 0.8rem;
  border-radius: 4px;
  cursor: pointer;
  border: 1px solid lightgray;
  color: gray;
}

.reserved-text {
  display: flex;
  gap: 0.4rem;
}

.reserved-popup {
  background-color: #0daa8b;
  height: 16px;
  padding: 0.2rem 0.3rem;
  border-radius: 4px;
}
