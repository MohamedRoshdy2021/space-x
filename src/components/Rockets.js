import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { reserveRocket, cancelReservation, fetchRocketsData } from '../Redux/Rockets/RocketsSlice';
import './Rockets.css';

function Rockets() {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.data);
  const reservedRockets = useSelector((state) => state.rockets.reservedRockets);

  useEffect(() => {
    dispatch(fetchRocketsData());
  }, [dispatch]);

  const toggleReservation = (rocketId) => {
    if (reservedRockets.includes(rocketId)) {
      dispatch(cancelReservation(rocketId));
    } else {
      dispatch(reserveRocket(rocketId));
    }
  };

  return (
    <div>
      {rockets.slice(0, 4).map((rocket) => (
        <div key={rocket.id} className="parent-div">
          <div>
            <img className="rockets-image" src={rocket.flickr_images} alt={rocket.name} />
          </div>
          <div className="text-div">
            <h2>{rocket.rocket_name}</h2>
            <div className="reserved-text">
              {reservedRockets.includes(rocket.id) && <p className="reserved-popup">Reserved</p>}
              <p>{rocket.description}</p>
            </div>
            {reservedRockets.includes(rocket.id) ? (
              <button
                type="button"
                className="cancel-reserve-button"
                onClick={() => toggleReservation(rocket.id)}
              >
                Cancel Reservation
              </button>
            ) : (
              <button
                type="button"
                className="reserve-button"
                onClick={() => toggleReservation(rocket.id)}
              >
                Reserve Rocket
              </button>
            )}
          </div>
        </div>
      ))}
    </div>
  );
}

export default Rockets;
