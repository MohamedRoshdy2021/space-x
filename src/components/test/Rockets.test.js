import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Rockets from '../Rockets'; // Make sure to use the correct import path

const mockStore = configureStore([thunk]);

test('matches snapshot for Rockets component', () => {
  const store = mockStore({
    rockets: {
      isLoading: false,
      data: [
        {
          id: 1,
          rocket_name: 'Rocket 1',
          description: 'Description 1',
          flickr_images: 'rocket1.jpg',
        },
        {
          id: 2,
          rocket_name: 'Rocket 2',
          description: 'Description 2',
          flickr_images: 'rocket2.jpg',
        },
      ],
      reservedRockets: [1],
    },
  });

  const { asFragment } = render(
    <Provider store={store}>
      <Rockets />
    </Provider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
