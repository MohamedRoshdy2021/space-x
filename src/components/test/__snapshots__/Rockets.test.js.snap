// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for Rockets component 1`] = `
<DocumentFragment>
  <div>
    <div
      class="parent-div"
    >
      <div>
        <img
          class="rockets-image"
          src="rocket1.jpg"
        />
      </div>
      <div
        class="text-div"
      >
        <h2>
          Rocket 1
        </h2>
        <div
          class="reserved-text"
        >
          <p
            class="reserved-popup"
          >
            Reserved
          </p>
          <p>
            Description 1
          </p>
        </div>
        <button
          class="cancel-reserve-button"
          type="button"
        >
          Cancel Reservation
        </button>
      </div>
    </div>
    <div
      class="parent-div"
    >
      <div>
        <img
          class="rockets-image"
          src="rocket2.jpg"
        />
      </div>
      <div
        class="text-div"
      >
        <h2>
          Rocket 2
        </h2>
        <div
          class="reserved-text"
        >
          <p>
            Description 2
          </p>
        </div>
        <button
          class="reserve-button"
          type="button"
        >
          Reserve Rocket
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
